### DISCUSSION TITLE
How to define CLI with subcommands and global arguments or flags using derive?

### DISCUSSION THREAD

**Discussion by mainrs** [pacak/bpaf#111] [Q&A] [answered]
I want a CLI that has a shared verbose flag for all subcommands. How can I define that using the derive API?

**Comment by pacak** [comment, marked as answer]
I'd have a top level Options struct containing all the shared flags as well as enum with commands:

```rust
#[derive(Debug, Clone, Bpaf)]
#[bpaf(options)]
struct Options {
    verbose: bool,
    #[bpaf(external)]
    action: Action
}

#[derive(Debug, Clone, Bpaf)]
enum Action {
    #[bpaf(command)]
    Foo,
    #[bpaf(command)]
    Bar,
}
```

**Reply by mainrs** [reply]
Is it possible to define a catch-all subcommand? My subcommands depend on the current folder structure. I have an `init` command that exists no matter what. But if `folderA` exists then there also has to exist a subcommand called `folder-a`.

**Reply by pacak** [reply]
Something like this would behave like a catch-all command: https://github.com/pacak/bpaf/blob/master/examples/fallback_command.rs

But this means doing things by hands. I'd have `folder-a` command present always and either check if it's valid with `guard` or do the same inside your app code.

**Reply by pacak** [reply]
You want to tell users what commands are available in `--help` like `git --help` prints everything even when not inside a git repo.

**Reply by mainrs** [reply]
The problem that I have is that the folder names can be arbitrary. It is out of my control.

**Reply by pacak** [reply]
So for every possible folder name you want to have a separate command? In that case one way would be
1. `fallback_command` example I gave above 
2. mentioning the fact that there are magical command names in doc comments for top level options
3. if options taken by inner commands are similar you can define a parser for that and call with `OptionParser::run_inner`, you can construct `Args` from `Vec<OsString>` you get.

**Reply by mainrs** [reply]
Thank you very much!!!

**Reply by pacak** [reply]
You should be able to hide call to the inner parser into `parse` parameter on `args` field so the end consumer will get a parsed set of parameters.
